File: assertion/c/question6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

_REPLACE_ME_WITH_DIRECTIVES_

int *calculateGrade(int size, int *input);

_REPLACE_ME_WITH_SOLUTION_

int *__workingAnswer(int size, int *input)
{
    for (int i = 0; i < size; i++)
    {
        int grade = input[i];
        if (grade >= 38 && grade % 5 != 0 && grade % 5 >= 3)
        {
            input[i] = grade + (5 - grade % 5);
        }
        else
        {
            input[i] = grade;
        }
    }

    return input;
}

typedef struct TestCase
{
    char *expected;
    char *got;
    char *arguments;
} TestCase;

// creates a random number between min and max
int __randomNumber(int min, int max)
{
    return (rand() % (max - min + 1)) + min;
}

int *genArray(int len)
{
    int *result = malloc(len * sizeof(int));
    for (int i = 0; i < len; i++)
    {
        result[i] = __randomNumber(10, 100);
    }
    return result;
}

// this is poorman's implementation
char *arrayToString(int len, int *arr)
{
    int index = 0;
    char *result = malloc(255 * sizeof(char));

    index += sprintf(result, "%c", '[');
    for (int i = 0; i < len - 1; i++)
    {
        index += sprintf(result + index, "%d, ", arr[i]);
    }
    sprintf(result + index, "%d%c", arr[len - 1], ']');

    return result;
}

int main()
{
    srand(time(0));

    TestCase testCases[10];

    int input[] = {73, 67, 38, 33};
    int answer[] = {75, 67, 40, 33};

    testCases[0].expected = arrayToString(4, answer);
    testCases[0].got = arrayToString(4, calculateGrade(4, input));
    testCases[0].arguments = "calculateGrade(4, [73, 67, 38, 33])";

    for (int i = 1; i < 10; i++)
    {
        TestCase *test = testCases + i;

        int len = __randomNumber(4, 20);

        int *input = genArray(len);
        char *args = arrayToString(len, input);
        int *input2 = malloc(len * sizeof(int));
        memcpy(input2, input, len * sizeof(int));

        int *expected = __workingAnswer(len, input);
        int *got = calculateGrade(len, input2);
        test->expected = arrayToString(len, expected);
        test->got = arrayToString(len, got);
        test->arguments = malloc(sizeof(char[250]));
        sprintf(test->arguments, "calculateGrade(%d, %s)", len, args);

        free(input);
        free(input2);
        free(args);
    }

    for (unsigned int i = 0; i < sizeof(testCases) / sizeof(TestCase); i++)
    {
        TestCase *test = testCases + i;

        if (strcmp(test->got, test->expected) == 0)
        {
            printf("# %d PASSING\n", i + 1);
        }
        else
        {
            printf("# %d FAILED\n", i + 1);
        }

        printf("> ARGUMENTS %s\n", test->arguments);
        printf("> EXPECTED %s\n", test->expected);
        printf("> GOT %s\n", test->got);

        free(test->expected);
        free(test->got);
        if (i >= 1)
        {
            free(test->arguments);
        }
    }
    return 0;
}

File: assertion/php/question6.php
<?php

_REPLACE_ME_WITH_DIRECTIVES_

_REPLACE_ME_WITH_SOLUTION_

function main(): void {
    $test_cases = [
        [
            "arguments" => "calculateGrade([73, 67, 38, 33])",
            "got" => calculateGrade([73, 67, 38, 33]),
            "expected" => [75, 67, 40, 33]
        ]
    ];

    function working_answer(array $arr): array {
        $out = [];
        for ($i = 0; $i < count($arr); $i++) {
            $grade = $arr[$i];

            if ($grade >= 38 && $grade % 5 !== 0 && $grade % 5 >= 3) {
                array_push($out, $grade + (5 - $grade % 5));
                continue;
            }
            array_push($out, $grade);
        }

        return $out;
    }

    for ($i = 0; $i < 9; $i++) {
        $input = [];
        $arr_len = rand(4, 20);

        for ($j = 0; $j < $arr_len; $j++) {
            array_push($input, rand(0, 100));
        }

        $expected = working_answer($input);
        $got = calculateGrade($input);
        $arguments = "calculateGrade(".json_encode($input).")";
        array_push($test_cases, [
            "got" => $got,
            "expected" => $expected,
            "arguments" => $arguments
        ]);
    }

    for ($i = 0; $i < count($test_cases); $i++) {
        $test = $test_cases[$i];
        
        if (is_null($test["got"])) {
            echo "# " . $i + 1 . " FAILED\n";
            echo "> ARGUMENTS " . $test["arguments"] . "\n";
            echo "> EXPECTED " . json_encode($test["expected"]) . "\n";
            echo "> GOT null\n";
            continue;
        }

        if (join(", ", $test["got"]) === join(", ", $test["expected"])) {
            echo "# ". $i + 1 . " PASSING\n";
            echo "> ARGUMENTS " . $test["arguments"] . "\n";
            echo "> EXPECTED " . json_encode($test["expected"]) . "\n";
            echo "> GOT " . json_encode($test["got"]) . "\n";
        } else {
            echo "# " . $i + 1 . " FAILED\n";
            echo "> ARGUMENTS " . $test["arguments"] . "\n";
            echo "> EXPECTED " . json_encode($test["expected"]) . "\n";
            echo "> GOT " . json_encode($test["got"]) . "\n";
        }
    }
}

main();